File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.js
/**
 * // Definition for a _Node.
 * function _Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {_Node} head
 * @return {_Node}
 */
var copyRandomList = function(head) {
    let ndeMap = new Map(), newHead = null;

    while (head) {
        let nde;
        if (!ndeMap.has(head)) {
            ndeMap.set(head, new Node(head.val))
        }
        nde = ndeMap.get(head);
        if (!newHead) {
            newHead = nde;
        }
        if (head.next && !ndeMap.has(head.next)) {
            ndeMap.set(head.next, new Node(head.next.val))

        }
        if (head.random && !ndeMap.has(head.random)) {
            ndeMap.set(head.random, new Node(head.random.val))
        }
        if (head.next) nde.next = ndeMap.get(head.next);
        if (head.random) nde.random = ndeMap.get(head.random);
        head = head.next;
    }
    return newHead;
};

File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.swift
/**
 * Definition for a Node.
 * public class Node {
 *     public var val: Int
 *     public var next: Node?
 *     public var random: Node?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *    	   self.random = nil
 *     }
 * }
 */

class Solution {
    func copyRandomList(_ head: Node?) -> Node? {
        var newHead: Node?, curr: Node? = head
        var map: [ObjectIdentifier: Node] = [:]
        while curr != nil {
            if map[ObjectIdentifier(curr!)] == nil {map[ObjectIdentifier(curr!)] = Node(curr!.val)}
            var currNode: Node = map[ObjectIdentifier(curr!)]!
            if newHead == nil { newHead = currNode }

            if var next: Node = curr?.next {
                if map[ObjectIdentifier(next)] == nil {map[ObjectIdentifier(next)] = Node(next.val)}
                currNode.next = map[ObjectIdentifier(next)]
            }
            if var random: Node = curr?.random {
                if map[ObjectIdentifier(random)] == nil {map[ObjectIdentifier(random)] = Node(random.val)}
                currNode.random = map[ObjectIdentifier(random)]
            }
            curr = curr?.next
        }
        return newHead
    }
}